### PR TITLE
Alternative approach to handling equality comparison mode for notifications 

### PR DESCRIPTION
**Pull request by kitchoi**
~This PR is currently a draft for discussion purposes.~

This PR presents an alternative approach to support `observe` even when the `List`, `Set` and `Dict` trait types have comparison mode set to equality.

The solution is to move the rich comparison from CTrait to the notifier, such that instead of silencing the notifications early in CTrait, the notifiers will do the equality check and silence the notification if applicable.

Pros:
- `ComparisonMode.equality` has the desired effect on trait notifications via both `on_trait_change` and `observe`, i.e. no need for #1122 

Cons:
- The equality check is done as many times as the number of notifiers attached to a trait. I hope this only presents a very slight performance cost.

Side-effect:
- Previously `post_setattr` (if defined) is not called if the notifications are silenced due to equality check. Now `post_setattr` will be called even if change notifications are silenced due to equality check.  The original behaviour looks to me an optimization that can be removed, but I am not entirely sure.

Uncertainty:
- It looks as though the code path that reaches rich comparison in `ctraits.c` only applies to trait types with `TraitType.trait`. e.g. notifications emitted for `Property` and `Delegate` do not go through these equality checks.  Filtering for `TraitType.trait` in the notifiers look enough, but I am not sure if there are other code paths using this rich comparison that have not been accounted for.

This approach may be an alternative to #1122.

If this works, it should present no visible changes to users of `traits`. ~However I think this is quite a substantial internal change that presents quite a bit of risk. Hence I'd be hesitant to let this into Traits 6.1 (it won't make it in anyway judging from the time line).  It is perhaps reasonable for `observe` to warn about equality mode and leave CTrait/`on_trait_change` untouched in at least one release.~ (Edit: 6.1 is now out 🎉 )

**Checklist**
- [x] Tests
- ~Update API reference (`docs/source/traits_api_reference`)~
- [x] Update User manual (`docs/source/traits_user_manual`)
- ~Update type annotation hints in `traits-stubs`~

### CODE REVIEW

File: traits/observation/_has_traits_helpers.py [hunk 1]
```diff
@@ -170,3 +113,23 @@ def observer_change_handler(event, graph, handler, target, dispatcher):
...
         )
+
+
+def ctrait_prevent_event(event):
+    """ Return true if the CTrait change event should be skipped.
+
+    Parameters
+    ----------
+    event : TraitChangeEvent
+    """
```
Comment (mdickinson) [inline traits/observation/_has_traits_helpers.py:124]: Nit: missing the "Returns" block in the docstring.

File: traits/observation/_has_traits_helpers.py [hunk 2]
```diff
@@ -170,3 +113,23 @@ def observer_change_handler(event, graph, handler, target, dispatcher):
...
+    if event.old is Uninitialized:
+        return True
+
+    ctrait = event.object.traits()[event.name]
+    if (ctrait.type == TraitKind.trait.name
+            and ctrait.comparison_mode == ComparisonMode.equality):
+        try:
+            return event.old == event.new
+        except Exception:
```
Comment (mdickinson) [inline traits/observation/_has_traits_helpers.py:133]: To keep the same semantics as in the original code, we probably want an extra `bool` call here: `return bool(event.old == event.new)`. The `bool` conversion is more likely to raise than the equality check - it should be very rare for an equality check to raise, but it's somewhat less rare for the interpretation of the result as a boolean to raise (e.g., NumPy arrays).

I think we should also be trying to get rid of this exception-swallowing behaviour at some point, but obviously not in this PR. Please could you open an issue?

(Same comments apply in `_change_accepted` below.)

File: traits/trait_notifiers.py [hunk 3]
```diff
@@ -628,3 +631,30 @@ class NewTraitChangeNotifyWrapper(TraitChangeNotifyWrapper):
...
+    ----------
+    object : HasTraits
+        The object on which the trait is changed.
+    name : str
+        The name of the trait changed.
+    old : any
+        The old value
+    new : any
+        The new value
+    """
```
Comment (mdickinson) [inline traits/trait_notifiers.py:649]: Nit: missing `Returns` block.

File: traits/observation/tests/test_has_traits_helpers.py [hunk 4]
```diff
@@ -137,159 +136,146 @@ def test_iter_objects_does_not_trigger_property(self):
...
     list_values = List(comparison_mode=2)
     dict_values = Dict(comparison_mode=2)
     set_values = Set(comparison_mode=2)
-    property_list = Property(List(comparison_mode=2))
-    container_in_union = Union(
-        None,
-        Set(comparison_mode=1),
-        comparison_mode=2,
-    )
+    number = Any(comparison_mode=2)
```
Comment (mdickinson) [inline traits/observation/tests/test_has_traits_helpers.py:151]: Off-topic for this PR, but we should change these to use the `ComparisonMode` constants.

Comment (mdickinson) [reply traits/observation/tests/test_has_traits_helpers.py:151]: Open #1229 for this, but I don't want to cause merge conflicts for this PR. So we should update #1229 after merging this PR.

Comment (kitchoi) [reply traits/observation/tests/test_has_traits_helpers.py:151]: OK!  I don't mind resolving the merge conflicts here either. Either way is fine for me.

File: traits/observation/_has_traits_helpers.py (hunk 2 above)
Comment (kitchoi) [reply traits/observation/_has_traits_helpers.py:133]: Sure I can open an issue, but actually I think it makes sense for the change handler to always fire if it fails to carry out an equality check....... maybe I should continue that discussion in that separate issue.

Comment (mdickinson) [reply traits/observation/_has_traits_helpers.py:133]: Maybe. My concern is the usual one with catching `Exception`: that exception may well represent a coding error, and we don't want that coding error to be silently hidden. So I want to see that exception propagated in most cases (or at the very least, logged).

Comment (kitchoi) [reply traits/observation/_has_traits_helpers.py:138]: Good point about the equality check could return something else that is not a bool. I changed that.

Comment (kitchoi) [reply traits/observation/_has_traits_helpers.py:138]: Opened and referenced #1230

File: traits/observation/_filtered_trait_observer.py [hunk 5]
```diff
@@ -11,14 +11,14 @@
 from traits.observation._has_traits_helpers import (
     iter_objects,
     observer_change_handler,
+    ctrait_prevent_event,
```
Comment (kitchoi) [inline traits/observation/_filtered_trait_observer.py:14]: Import abc...

**Comment by kitchoi** [comment]
Updated main comment: User manual is also updated now.

File: docs/source/traits_user_manual/notification.rst [hunk 6]
```diff
@@ -491,6 +491,10 @@ with |@observe| is discouraged as this special trait may be removed when
 Identity comparison mode for container traits
 `````````````````````````````````````````````
 
+.. note:: This section is for Traits 6.1 only. Since Traits 6.2, the
```
Comment (mdickinson) [inline docs/source/traits_user_manual/notification.rst:494]: I'd suggest removing the relevant documentation instead. This version of the docs (in master) isn't going to see the light of day before we release Traits 6.2, and when we do release Traits 6.2 the documentation for 6.1 should still be available. So these docs are specific for 6.2

File: traits/observation/_has_traits_helpers.py [hunk 7]
```diff
@@ -170,3 +113,30 @@ def observer_change_handler(event, graph, handler, target, dispatcher):
...
+def ctrait_prevent_event(event):
+    """ Return true if the CTrait change event should be skipped.
+
+    Parameters
+    ----------
+    event : TraitChangeEvent
+
+    Returns
+    -------
+    skipped: bool
```
Comment (mdickinson) [inline traits/observation/_has_traits_helpers.py:127]: reST / numpydoc nit:
```suggestion
    skipped : bool
```

File: traits/trait_notifiers.py [hunk 8]
```diff
@@ -628,3 +631,37 @@ class NewTraitChangeNotifyWrapper(TraitChangeNotifyWrapper):
...
+    name : str
+        The name of the trait changed.
+    old : any
+        The old value
+    new : any
+        The new value
+
+    Returns
+    -------
+    accepted: bool
```
Comment (mdickinson) [inline traits/trait_notifiers.py:652]: markup nitpick:
```suggestion
    accepted : bool
```

File: docs/source/traits_user_manual/notification.rst (hunk 6 above)
Comment (kitchoi) [reply docs/source/traits_user_manual/notification.rst:494]: Right, forgot we now have all the versions available online!

**Comment by mdickinson** [comment]
> I think we simply want to update the documentation to match the Traits 6.2 behaviour rather than describing what happened in 6.1.

We could possibly make use of the `.. versionchanged:: ...` markup to add a note about the change from 6.1 to 6.2, though.

File: docs/source/traits_user_manual/notification.rst (hunk 6 above)
Comment (mdickinson) [reply docs/source/traits_user_manual/notification.rst]: > Right, forgot we now have all the versions available online!

We do still need to figure out how to make all those documentation versions easy to find for the user, though. (But that's a separate issue.)

**Comment by kitchoi** [comment]
> We could possibly make use of the .. versionchanged:: ... markup to add a note about the change from 6.1 to 6.2, though.

I have removed the section. Maybe the change can later live in the changelog / release notes?

**Pull request merged**